asapdiscovery-data/asapdiscovery/data/selectors/mcs_selector.py

```diff
@@ -0,0 +1,128 @@
+from typing import Literal, Union
+
+import numpy as np
+from asapdiscovery.data.openeye import oechem
+from asapdiscovery.data.schema_v2.complex import Complex, ComplexBase, PreppedComplex
+from asapdiscovery.data.schema_v2.ligand import Ligand
+from asapdiscovery.data.schema_v2.pairs import CompoundStructurePair, DockingInputPair
+from asapdiscovery.data.selectors.selector import SelectorBase
+from pydantic import Field
+
+
+class MCSSelector(SelectorBase):
+    """
+    Selects ligand and complex pairs based on a maximum common substructure
+    (MCS) search.
+    """
+
+    expander_type: Literal["PairwiseSelector"] = "PairwiseSelector"
+
+    structure_based: bool = Field(
+        False,
+        description="Whether to use a structure-based search (True) or a more strict element-based search (False).",
+    )
+
+    def _select(
+        self,
+        ligands: list[Ligand],
+        complexes: list[Union[Complex, PreppedComplex]],
+        n_select: int = 1,
+    ) -> list[Union[CompoundStructurePair, DockingInputPair]]:
+        """
+        Selects ligand and complex pairs based on maximum common substructure
+        (MCS) search.
+
+        Parameters
+        ----------
+        ligands : list[Ligand]
+            List of ligands to search for
+        complexes : list[Union[Complex, PreppedComplex]]]
+            List of complexes to search in
+        n_select : int, optional
+            Draw top n_select matched molecules for each ligand (default: 1) this means that the
+            number of pairs returned is n_select * len(ligands)
+
+        Returns
+        -------
+        list[tuple[Ligand, Complex]]
+            List of ligand and complex pairs
+        """
+
+        if not all(isinstance(c, ComplexBase) for c in complexes):
+            raise ValueError("All complexes must be of the same type")
+
+        pair_cls = self._pair_type_from_complex(complexes[0])
```

Comment:
maybe add some checking here that all `complexes` are the same class? might help prevent some weird errors down the line